### PR TITLE
[PVM, DAC] General proposal

### PR DESCRIPTION
**Pull request by evlekht**
## Why this should be merged
This PR adds new proposal type: generalProposal. This proposal can have up to 3 options, proposal duration is between 1 day and 1 month and proposal has configurable mostVotedThreshold ("qualified majority") and totalVotedThreshold ("quorum").
Proposal options consists of arbitrary bytes, length is limited to 256 bytes. Regardless of proposal outcome, it doesn't affect state in any way.
## How this works
Adds new proposal type and corresponding logic utilizing previously added dac-proposals feature.
## How this was tested
Unit-tests

### CODE REVIEW

File: vms/platformvm/dac/camino_general_proposal.go [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+	remainingVotes := p.TotalAllowedVoters - voted
+
+	secondMostVotedWeight := uint32(0)
+	if len(p.Options) > 1 {
+		for index, option := range p.Options {
+			if option.Weight > secondMostVotedWeight && index != int(mostVotedIndex) {
+				secondMostVotedWeight = option.Weight
+			}
+		}
+	}
```
Comment (knikos) [inline vms/platformvm/dac/camino_general_proposal.go:188]: i would argue that having a sort-by-weight logic on p.Options would be much more intuitive and easy to understand

File: vms/platformvm/dac/camino_general_proposal.go [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+
+	// mostVotedTheshold = voted * p.MostVotedThresholdNominator / fractionDenominatorBig
+	mostVotedThresholdNominator := big.NewInt(int64(p.MostVotedThresholdNominator))
+	mostVotedThesholdBig := big.NewInt(int64(voted))
+	mostVotedThesholdBig.Mul(mostVotedThesholdBig, mostVotedThresholdNominator)
+	mostVotedThesholdBig.Div(mostVotedThesholdBig, fractionDenominatorBig)
+	mostVotedTheshold := uint32(mostVotedThesholdBig.Uint64())
```
Comment (knikos) [inline vms/platformvm/dac/camino_general_proposal.go]: typo -> missing `r` in Threshold

File: vms/platformvm/dac/camino_general_proposal.go [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+		for index, option := range p.Options {
+			if option.Weight > secondMostVotedWeight && index != int(mostVotedIndex) {
+				secondMostVotedWeight = option.Weight
+			}
+		}
+	}
+
+	// mostVotedTheshold = voted * p.MostVotedThresholdNominator / fractionDenominatorBig
+	mostVotedThresholdNominator := big.NewInt(int64(p.MostVotedThresholdNominator))
+	mostVotedThesholdBig := big.NewInt(int64(voted))
```
Comment (knikos) [inline vms/platformvm/dac/camino_general_proposal.go]: typo -> missing `r` in Threshold

File: vms/platformvm/txs/executor/dac/camino_dac.go [hunk 4]
```diff
@@ -349,6 +349,43 @@ func (e *proposalBondTxIDsGetter) ExcludeMemberProposal(proposal *dac.ExcludeMem
...
+	proposerNodeID := ids.NodeID(proposerNodeShortID)
+
+	if _, err = e.state.GetCurrentValidator(constants.PrimaryNetworkID, proposerNodeID); err == database.ErrNotFound {
+		return errNoActiveValidator
+	}
+	return err
+}
+
+func (*proposalExecutor) GeneralProposal(*dac.GeneralProposalState) error {
+	return nil
```
Comment (knikos) [inline vms/platformvm/txs/executor/dac/camino_dac.go:353]: what does this mean? are we expecting future work here? could you note it if so?

File: vms/platformvm/txs/executor/dac/camino_dac.go [hunk 5]
```diff
@@ -349,6 +349,43 @@ func (e *proposalBondTxIDsGetter) ExcludeMemberProposal(proposal *dac.ExcludeMem
...
+		return errNoActiveValidator
+	case err != nil:
+		return err
+	}
+	proposerNodeID := ids.NodeID(proposerNodeShortID)
+
+	if _, err = e.state.GetCurrentValidator(constants.PrimaryNetworkID, proposerNodeID); err == database.ErrNotFound {
+		return errNoActiveValidator
+	}
+	return err
```
Comment (knikos) [inline vms/platformvm/txs/executor/dac/camino_dac.go]: similar logic seems a bit more granular in ExcludeMemberProposal (`vms/platformvm/txs/executor/dac/camino_dac.go`). 
I see a need for repetition of this logic and thus suggest to extract it as a function that does the following

```go
func (cs *caminoState) hasActiveValidator(.....) (bool,error) {
// get proposer nodeID
		proposerNodeShortID, err := e.state.GetShortIDLink(e.addProposalTx.ProposerAddress, state.ShortLinkKeyRegisterNode)
		switch {
		case err == database.ErrNotFound:
			return errNoActiveValidator
		case err != nil:
			return err
		}
		proposerNodeID := ids.NodeID(proposerNodeShortID)

		// get proposer active validator
		_, err = e.state.GetCurrentValidator(constants.PrimaryNetworkID, proposerNodeID)
		switch {
		case err == database.ErrNotFound:
			return errNoActiveValidator
		case err != nil:
			return err
		}
```

File: vms/platformvm/dac/camino_general_proposal_test.go [hunk 6]
```diff
@@ -0,0 +1,916 @@
...
+			require.Equal(t, tt.expectedProposal, tt.proposal)
+			require.Equal(t, tt.expectedProposalState, proposalState)
+		})
+	}
+}
+
+func TestGeneralProposalStateAddVote(t *testing.T) {
+	voterAddr1 := ids.ShortID{1}
+	voterAddr2 := ids.ShortID{1}
+	voterAddr3 := ids.ShortID{1}
```
Comment (knikos) [inline vms/platformvm/dac/camino_general_proposal_test.go]: here we have this accidental copy-paste of all addresses have same short IDa

File: vms/platformvm/txs/executor/dac/camino_dac.go (hunk 4 above)
Comment (evlekht) [reply vms/platformvm/txs/executor/dac/camino_dac.go:353]: no, it explicitly mentioned in PR description, that proposal won't have any effect on blockchain

File: vms/platformvm/dac/camino_general_proposal.go (hunk 1 above)
Comment (evlekht) [reply vms/platformvm/dac/camino_general_proposal.go:188]: no, options must retain the same indexes - thats what identifies them

Comment (knikos) [reply vms/platformvm/dac/camino_general_proposal.go:188]: you could perform it on a copy.

Comment (evlekht) [reply vms/platformvm/dac/camino_general_proposal.go:188]: honestly, don't know
doesn't look as its easier to grasp to me - both has some "drawbacks" :)
```go
	secondMostVotedWeight := uint32(0)
	if len(p.Options) > 1 {
		for index, option := range p.Options {
			if option.Weight > secondMostVotedWeight && index != int(mostVotedIndex) {
				secondMostVotedWeight = option.Weight
			}
		}
	}
```
vs
```go
	secondMostVotedWeight := uint32(0)
	if len(p.Options) > 1 {
		var options []SimpleVoteOption[[]byte]
		copy(p.Options, options)
		slices.SortFunc(options, func(a, b SimpleVoteOption[[]byte]) bool { return a.Weight < b.Weight })
		secondMostVotedWeight = options[len(options)-2].Weight
	}
```

**Review by knikos — DISMISSED** [review]

**Review by knikos — APPROVED** [review]

**Pull request merged**